src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/CosmosDb/CosmosCommon.cs

```diff
@@ -0,0 +1,174 @@
+
+        public const string IntegrationName = nameof(IntegrationIds.CosmosDb);
+
+        private const string OperationName = "cosmosdb.query";
+        private const string ServiceName = "cosmosdb";
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(CosmosCommon));
+
+        public static CallTargetState CreateContainerCallStateExt<TTarget, TQueryDefinition>(TTarget instance, TQueryDefinition queryDefinition)
+        {
+            return CreateCosmosDbCallState(GetContainerPropterties, instance, queryDefinition);
+        }
+
+        public static CallTargetState CreateDatabaseCallStateExt<TTarget, TQueryDefinition>(TTarget instance, TQueryDefinition queryDefinition)
+        {
+            return CreateCosmosDbCallState(GetDatabasePropterties, instance, queryDefinition);
+        }
+
+        public static CallTargetState CreateClientCallStateExt<TTarget, TQueryDefinition>(TTarget instance, TQueryDefinition queryDefinition)
+        {
+            return CreateCosmosDbCallState(GetClientPropterties, instance, queryDefinition);
+        }
+
+        private static CallTargetState CreateCosmosDbCallState<TTarget, TQueryDefinition>(Func<object, Tuple<string, string, string>> extractProperties, TTarget instance, TQueryDefinition queryDefinition)
+        {
+            if (!Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationName))
+            {
+                // integration disabled, don't create a scope, skip this trace
+                return CallTargetState.GetDefault();
+            }
+
+            try
+            {
+                string query;
+                if (queryDefinition is string queryDefinitionString)
+                {
+                    query = queryDefinitionString;
+                }
+                else
+                {
+                    var success = queryDefinition.TryDuckCast<QueryDefinitionStruct>(out var queryDefinitionObj);
+                    query = queryDefinitionObj.QueryText;
+                }
+
+                var tracer = Tracer.Instance;
+
+                var parent = tracer.ActiveScope?.Span;
+
+                if (parent != null &&
+                    parent.Type == SpanTypes.Sql &&
+                    parent.GetTag(Tags.DbType) == "cosmosdb" &&
+                    parent.ResourceName == query)
+                {
+                    // we are already instrumenting this,
+                    // don't instrument nested methods that belong to the same stacktrace
+                    return new CallTargetState(null);
+                }
+
+                var properties = extractProperties(instance);
+
+                string containerId = properties.Item1;
+                string databaseId = properties.Item2;
+                string endpoint = properties.Item3;
+
+                if (endpoint == null)
+                {
+                    Console.WriteLine();
+                }
+
+                var tags = new CosmosDbTags
```

Comment:
We should call `InstrumentationTags.SetAnalyticsSampleRate()` from every integration (with `enabledWithGlobalSetting: false` in this case). See an example here:
https://github.com/DataDog/dd-trace-dotnet/blob/b68cdfef714b378cd61ced1fa2894e0e215d90a9/src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/GraphQL/GraphQLCommon.cs#L52

Reply:
I have add this, but isn't ensure this is called from every single integration a bit error prone? Could they be parameters in the base constructor tags for example?

Reply:
Yes, we should look at improving the dev experience here. It's gone through several iterations over time and we haven't gone back to clean it up 😅